nomenclature/processor/generic.py

```diff
@@ -0,0 +1,153 @@
+
+
+class AggregationItem(BaseModel):
+    """Item used for aggregation-mapping"""
+
+    name: str
+    components: list[str]
+
+
+class Aggregator(Processor):
+    """Aggregation or renaming of an IamDataFrame on a `dimension`"""
+    file: FilePath
+    dimension: str
+    mapping: list[AggregationItem]
+
+    def apply(self, df: IamDataFrame) -> IamDataFrame:
+        """Apply region processing
+
+        Parameters
+        ----------
+        df : IamDataFrame
+            Input data that the region processing is applied to
+
+        Returns
+        -------
+        IamDataFrame:
+            Processed data
+
+        """
+        return df.rename(
+            mapping={self.dimension: self.rename_mapping},
+            check_duplicates=False,
+        )
+
+    @property
+    def rename_mapping(self):
+        rename_dict = {}
+
+        for item in self.mapping:
+            for c in item.components:
+                rename_dict[c] = item.name
+
+        return rename_dict
+
+    @field_validator("mapping")
+    def validate_target_names(cls, v, info: ValidationInfo):
+        _validate_items([item.name for item in v], info, "Duplicate target")
+        return v
+
+    @field_validator("mapping")
+    def validate_components(cls, v, info: ValidationInfo):
+        # components have to be unique for creating rename-mapping (component -> target)
+        all_components = list()
+        for item in v:
+            all_components.extend(item.components)
+        _validate_items(all_components, info, "Duplicate component")
+        return v
+
+    @field_validator("mapping")
+    def validate_target_vs_components(cls, v, info: ValidationInfo):
+        # guard against having identical target and component
+        _codes = list()
+        for item in v:
+            _codes.append(item.name)
+            _codes.extend(item.components)
+        _validate_items(_codes, info, "Non-unique target and component")
+        return v
+
+    @property
+    def codes(self):
+        _codes = list()
+        for item in self.mapping:
+            _codes.append(item.name)
+            _codes.extend(item.components)
+        return _codes
+
+    def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
+        error = None
+        # check for codes that are not defined in the codelists
+        codelist = getattr(dsd, self.dimension, None)
+        # no validation if codelist is not defined or filter-item is None
+        if codelist is None:
+            error = f"Dimension '{self.dimension}' not found in DataStructureDefinition"
+        elif invalid := codelist.validate_items(self.codes):
+            error = (
+                f"The following {self.dimension}s are not defined in the "
+                "DataStructureDefinition:\n - " + "\n - ".join(invalid)
+            )
+        if error:
+            raise ValueError(error + "\nin " + str(self.file) + "")
+
+    @classmethod
+    def from_file(cls, file: Path | str):
```

Comment:
If you rename the `mapping` attribute to `aggregate` and use `dict[str, list[str]]` in favor of `AggregationItem`, I think you should be able to remove this function almost entirely.

Reply:
Yes, exactly